clippy_lints/src/redundant_test_prefix.rs

```diff
@@ -0,0 +1,143 @@
+use clippy_config::Conf;
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::visitors::for_each_expr;
+use clippy_utils::{is_in_cfg_test, is_test_function};
+use rustc_errors::Applicability;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{self as hir, Body, ExprKind, FnDecl};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::Span;
+use rustc_span::def_id::LocalDefId;
+use std::ops::ControlFlow;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for test functions (functions annotated with `#[test]`) that are prefixed
+    /// with `test_` which is redundant.
+    ///
+    /// ### Why is this bad?
+    /// This is redundant because test functions are already annotated with `#[test]`.
+    /// Moreover, it clutters the output of `cargo test` since test functions are expanded as
+    /// `module::tests::test_use_case` in the output. Without the redundant prefix, the output
+    /// becomes `module::tests::use_case`, which is more readable.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// #[test]
+    /// fn test_use_case() {
+    ///     // test code
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// fn use_case() {
+    ///    // test code
+    /// }
+    /// ```
+    #[clippy::version = "1.84.0"]
+    pub REDUNDANT_TEST_PREFIX,
+    pedantic,
+    "redundant `test_` prefix in test function name"
+}
+
+pub struct RedundantTestPrefix {
+    check_outside_test_cfg: bool,
+    custom_sufix: String,
+}
+
+impl RedundantTestPrefix {
+    pub fn new(conf: &'static Conf) -> Self {
+        Self {
+            check_outside_test_cfg: conf.redundant_test_prefix_check_outside_cfg_test,
+            custom_sufix: conf.redundant_test_prefix_custom_suffix.clone(),
+        }
+    }
+}
+
+impl_lint_pass!(RedundantTestPrefix => [REDUNDANT_TEST_PREFIX]);
+
+impl<'tcx> LateLintPass<'tcx> for RedundantTestPrefix {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        kind: FnKind<'_>,
+        _decl: &FnDecl<'_>,
+        body: &'tcx Body<'_>,
+        _span: Span,
+        fn_def_id: LocalDefId,
+    ) {
+        // Ignore methods and closures.
+        let FnKind::ItemFn(ref ident, ..) = kind else {
+            return;
+        };
+
+        // Skip the lint if the function is not within a node marked with `#[cfg(test)]` attribute.
+        // Continue if the function is inside the node marked with `#[cfg(test)]` or lint is enforced
+        // via configuration (most likely to include integration tests in lint's scope).
+        if !(self.check_outside_test_cfg || is_in_cfg_test(cx.tcx, body.id().hir_id)) {
+            return;
+        }
+
+        if is_test_function(cx.tcx, cx.tcx.local_def_id_to_hir_id(fn_def_id)) && ident.as_str().starts_with("test_") {
+            let mut non_prefixed = ident.as_str().trim_start_matches("test_").to_string();
```

Comment:
In theory, it can happen that the new string can no longer be an ident. For example, if it starts with an integer or is empty. IDK, if there is a simple way to check this :thinking: 

Examples would be `test_1`, `test_const`, `test_`.

It looks like the [lexer](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lexer/cursor/struct.Cursor.html#method.invalid_ident) is responsible for checking this in rustc, but it doesn't have an easily accessible util function. We should be able to reuse the lexer. (rn I just don't know how from the top of my head)

Reply:
Nice catch! I will look into it, starting from sample case and investigate how we can resolve this.